packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -104,102 +136,209 @@ class MouseTracker extends ChangeNotifier {
+    for (final MouseTrackerAnnotation annotation in enteringAnnotations) {
+      if (annotation.onEnter != null) {
+        annotation.onEnter(PointerEnterEvent.fromMouseEvent(mostRecentEvent));
+      }
     }
-    if (event is PointerRemovedEvent) {
-      _removeMouseEvent(deviceId, event);
-      // If the mouse was removed, then we need to schedule one more check to
-      // exit any annotations that were active.
-      _sendMouseNotifications(<int>{deviceId});
-    } else {
-      if (event is PointerMoveEvent || event is PointerHoverEvent || event is PointerDownEvent) {
-        final PointerEvent lastEvent = _lastMouseEvent[deviceId];
-        _addMouseEvent(deviceId, event);
-        if (lastEvent == null ||
-            lastEvent is PointerAddedEvent || lastEvent.position != event.position) {
-          // Only schedule a frame if we have our first event, or if the
-          // location of the mouse has changed, and only if there are tracked annotations.
-          _sendMouseNotifications(<int>{deviceId});
+
+    // Send hover events in reverse visual order.
+    // No solid reasons have been brough up on the order between the hover
+    // events, except for keeping it aligned with enter events for simplicity.
+    if (mostRecentEvent is PointerHoverEvent) {
+      final Iterable<MouseTrackerAnnotation> hoveringAnnotations =
+        nextAnnotations.toList().reversed;
+      for (final MouseTrackerAnnotation annotation in hoveringAnnotations) {
+        if (annotation.onHover != null) {
+          annotation.onHover(mostRecentEvent);
         }
       }
     }
   }
 
-  _TrackedAnnotation _findAnnotation(MouseTrackerAnnotation annotation) {
-    final _TrackedAnnotation trackedAnnotation = _trackedAnnotations[annotation];
-    assert(
-        trackedAnnotation != null,
-        'Unable to find annotation $annotation in tracked annotations. '
-        'Check that attachAnnotation has been called for all annotated layers.');
-    return trackedAnnotation;
+  // Collect the latest states of the given mouse device `device`, and call
+  // interested callbacks.
+  //
+  // The enter or exit events are called for annotations that the pointer
+  // enters or leaves, while hover events are always called for each
+  // annotations that the pointer stays in, even if the pointer has not moved
+  // since the last call. Therefore it's caller's responsibility to check if
+  // the pointer has moved.
+  //
+  // If `disconnectedMouseState` is provided, this state will be used instead,
+  // but this mouse will be hovering no annotations.
+  void _checkDeviceUpdates({
+    int device,
+    _MouseState disconnectedMouseState,
+  }) {
+    final _MouseState mouseState = disconnectedMouseState ?? _mouseStates[device];
+    final bool thisDeviceIsConnected = mouseState != disconnectedMouseState;
+    assert(mouseState != null);
+
+    final LinkedHashSet<MouseTrackerAnnotation> nextAnnotations =
```

Comment:
This block is awkwardly formatted. The lines with the ? and : need to be indented more, I think?